### PR TITLE
fix: prevent SQL Browser DAC response parsing panic (DoS)

### PR DESCRIPTION
**Pull request by saurabh500**
## Summary

Fuzzing the SQL Server Browser response parsers turned up a crash in `parseDAC`, the parser for the Dedicated Admin Connection (DAC) reply on UDP port 1434.

`parseDAC` read the TCP port from the wrong offset. A `CLNT_UCAST_DAC` reply is exactly 6 bytes ([MS-SQLR 2.2.7](https://learn.microsoft.com/en-us/openspecs/windows_protocols/mc-sqlr/1ea6e25f-bff9-4364-ba21-5dc449a601b7)): `SVR_RESP` (`0x05`), a 2-byte length, a protocol version byte, then the little-endian port at offset 4. The code sliced `msg[5:]`:

```go
if len(msg) == 6 && msg[0] == 5 {
    results[strings.ToUpper(instance)]["tcp"] = fmt.Sprint(binary.LittleEndian.Uint16(msg[5:]))
}
```

Since the guard is `len(msg) == 6`, `msg[5:]` is always a 1-byte slice and `Uint16` needs 2 — so this panicked on *every* well-formed reply:

```
panic: runtime error: index out of range [1] with length 1
encoding/binary.littleEndian.Uint16(...)
github.com/microsoft/go-mssqldb.parseDAC(...)   tds.go:26
github.com/microsoft/go-mssqldb.getInstances(...) tds.go:91
github.com/microsoft/go-mssqldb.dialConnection(...) tds.go:968
```

The same line also indexes `results[...]` before that key exists, yielding a `nil` `map[string]string` and writing into it. Argument evaluation ordering meant the slice panic fired first and masked it, so fixing only the offset would have converted one panic into another (`assignment to entry in nil map`).

Both run inside `connect()` with no `recover()` anywhere in the call stack, so the panic takes down the whole process instead of failing the connection attempt.

## Fix

```go
if len(msg) == 6 && msg[0] == 5 {
    name := strings.ToUpper(instance)
    results[name] = map[string]string{
        "InstanceName": name,
        "tcp":          fmt.Sprint(binary.LittleEndian.Uint16(msg[4:6])),
    }
}
```

Read from `msg[4:6]` and build the inner map before writing.

### Second defect the crash was hiding

`parseDAC` never set an `InstanceName` key, but `tcpDialer.ParseBrowserData` matches browser entries on exactly that key:

```go
inst, ok = i["InstanceName"]
if ok && stringForInstanceNameComparison(inst) == instanceName {
    strport, ok = i["tcp"]
```

With only `"tcp"` present the lookup always missed, so every DAC connection would have failed with `no instance matching '%v' returned from host '%v'`. The panic fired first, so this was never observable. Fixing the crash without this would have left the feature non-functional, so `InstanceName` is now populated and covered by a test that drives `parseDAC` output through `ParseBrowserData`.

## Tests

- **`FuzzParseDAC`** / **`FuzzParseInstances`** (new) — native Go fuzz targets for both Browser response parsers. The reply arrives over UDP and is fully attacker-controllable, so the property under test is that parsing never panics. `FuzzParseDAC` reproduces the crash directly from its seed corpus.
- **`TestParseDAC`** (updated) — the existing cases built 7-byte messages, so they never entered the `len(msg) == 6` branch and passed against the broken code. They now use spec-shaped 6-byte replies and assert the decoded port instead of only checking for absence of a panic.
- **`TestParseDACParseBrowserData`** (new) — end-to-end check that `parseDAC` output resolves to `p.Port == 1434` through `ParseBrowserData`.

Before the fix:

```
--- FAIL: FuzzParseDAC (0.94s)
    testing.go:1825: panic: runtime error: index out of range [1] with length 1
        github.com/microsoft/go-mssqldb.parseDAC(...)  tds.go:26
    failure while testing seed corpus entry: FuzzParseDAC/seed#0
```

After: both fuzz targets run clean (~3.8M and ~2.7M execs, no crashes), `go vet` is clean, and `go test .` passes.

### CODE REVIEW

File: tds.go [hunk 1]
```diff
@@ -20,10 +20,18 @@ import (
...
+// PROTOCOLVERSION byte, then the little-endian TCP_DAC_PORT at offset 4. The
+// message comes from an untrusted UDP source, so it must never panic.
 func parseDAC(msg []byte, instance string) msdsn.BrowserData {
 	results := msdsn.BrowserData{}
 	if len(msg) == 6 && msg[0] == 5 {
-		results[strings.ToUpper(instance)]["tcp"] = fmt.Sprint(binary.LittleEndian.Uint16(msg[5:]))
+		name := strings.ToUpper(instance)
+		results[name] = map[string]string{
+			"InstanceName": name,
+			"tcp":          fmt.Sprint(binary.LittleEndian.Uint16(msg[4:6])),
```
[the remark this answers is not in the document]
Comment (saurabh500) [reply tds.go]: Good catch — fixed in 3e9b21c.

`parseDAC` now checks both fixed fields before exposing the port:

```go
if len(msg) == 6 && msg[0] == 5 && binary.LittleEndian.Uint16(msg[1:3]) == 6 && msg[3] == 1 {
```

I checked this against MC-SQLR before tightening it, since rejecting replies the driver previously accepted is a behavior change:

- **2.2.6** fixes both fields for the DAC response — `RESP_SIZE` MUST be `0x0006` (the length of the whole packet, which overrides the "subsequent RESP_DATA" definition in 2.2.5) and `PROTOCOLVERSION` MUST be `0x01`. Only `TCP_DAC_PORT` is variable.
- **Appendix A** records no version-specific deviation for 2.2.6 across SQL Server 2000 through 2025 — the only message-syntax footnote is `<1>`, against 2.2.5. So no conforming Browser reply changes behavior here.
- **3.2.2 `<2>`** describes the reference client doing the same thing: for `CLNT_UCAST_DAC`, a valid response is passed to the higher layer and "if the response is not valid, an error is passed to the higher layer."

A rejected reply now yields empty `BrowserData`, which surfaces through `ParseBrowserData` as the normal `no instance matching ...` connection error rather than a silently-followed port.

Added `wrong resp size` and `wrong protocol version` cases to `TestParseDAC` asserting no entry is produced, and kept the inconsistent-header fuzz seeds — they now exercise the rejection path instead of acceptance. `FuzzParseDAC` is clean over ~1M execs.

One scoping note on the security framing: this is correctness hygiene rather than a security boundary. An attacker who can win the UDP race can just send a well-formed reply carrying a malicious port, so validation doesn't stop redirection by a capable attacker — the crash fix is the part that closes the DoS. Still worth rejecting malformed input, which is why I took the change.


> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Comment by saurabh500** [comment]
> `TdsBuffer_Write_Large-4    122.3n ± 0%   182.8n ± 1%   +49.49% (p=0.000 n=10)`

This one is bimodal benchmark noise rather than a regression from this PR — the same job reported the opposite sign on its previous run:

| run | main | PR | reported |
| --- | --- | --- | --- |
| earlier (merge `a1f4a80`) | 182.8n | 121.5n | −33.56% |
| latest (merge `0af87c1`) | 122.3n | 182.8n | +49.49% |

The branch head was unchanged at `3e9b21c` across both runs — only `main` advanced, so the PR-side code was byte-identical — yet its measurement moved 121.5n → 182.8n while the main-side moved the other way. Both arms visited both levels, so `TdsBuffer_Write_Large` is settling into one of two states (~122n / ~183n) per run instead of tracking the diff.

There is no mechanism for it either: this PR only changes `parseDAC` and its tests, and `TdsBuffer_Write_Large` never calls that path. Every other benchmark is within ±3%, and the job's own "Check for regressions" step passed.

**Comment by saurabh500** [comment]
> `getInstances` allocates only a six-byte DAC receive buffer (`tds.go:80`), so UDP `Read` truncates an oversized datagram before passing it here. A packet with a valid six-byte prefix plus trailing bytes therefore still satisfies `len(msg) == 6`, contrary to this exact-length validation and the new "too long" test.

Good catch, and confirmed on Linux. Fixed in 49aa148.

I verified it rather than taking it on faith, by cross-compiling the test binary and running it under a real Linux kernel. With the old 6-byte buffer, a 9-byte reply whose first 6 bytes are well-formed is silently truncated and **accepted**:

```
--- FAIL: TestGetInstancesDACResponseSize/oversized_reply_with_valid_prefix
    Error: Should be empty, but was map[MYINSTANCE:map[InstanceName:MYINSTANCE tcp:1434]]
    Messages: expected the reply to be rejected
```

So the exact-length check really was unreachable through the caller on truncating platforms.

The fix reads into a 7-byte buffer, which keeps an oversized datagram observable so `parseDAC` discards it. Worth noting the platform split I found while testing: Windows does not truncate, it fails the read with `WSAEMSGSIZE`, so oversized replies were already rejected there. The new `TestGetInstancesDACResponseSize` covers the real UDP read path over a loopback socket and asserts the platform-independent invariant — an oversized reply never resolves a DAC port — so it passes on both.

On the suggested `ReadMsgUDP` alternative: I went with the larger buffer instead, since the connection here comes back from the `Dialer` interface as a `net.Conn`. Reaching for `ReadMsgUDP` would mean type-asserting to `*net.UDPConn`, which breaks for custom dialers, for no detection benefit over the extra byte.

As before, this is correctness hygiene rather than a security boundary: anyone able to win the UDP race can simply send an exactly-6-byte well-formed reply.

File: tds.go [hunk 2]
```diff
@@ -66,7 +77,11 @@ func getInstances(ctx context.Context, d Dialer, address string, browserMsg msds
 		bmsg[0] = byte(msdsn.BrowserDAC)
 		bmsg[1] = 1
 		_ = copy(bmsg[3:], instance)
```
Comment (chlafreniere) [inline tds.go]: **AI-generated review finding (Medium):** The DAC request encodes the instance name at the wrong offset. `copy(bmsg[3:], instance)` leaves byte 2 as NUL, producing `0F 01 00 <instance>` rather than the MC-SQLR layout `0F 01 <instance> 00`. A named-instance DAC lookup can therefore fail or be interpreted as the default instance, potentially connecting to the wrong instance. Please copy into `bmsg[2:]` so the zero-initialized final byte remains the terminator, and add a request-capture test for the exact packet bytes.

File: tds.go [hunk 3]
```diff
@@ -20,10 +20,21 @@ import (
...
+// so this must never panic.
 func parseDAC(msg []byte, instance string) msdsn.BrowserData {
 	results := msdsn.BrowserData{}
-	if len(msg) == 6 && msg[0] == 5 {
-		results[strings.ToUpper(instance)]["tcp"] = fmt.Sprint(binary.LittleEndian.Uint16(msg[5:]))
+	if len(msg) == 6 && msg[0] == 5 && binary.LittleEndian.Uint16(msg[1:3]) == 6 && msg[3] == 1 {
+		name := strings.ToUpper(instance)
+		results[name] = map[string]string{
+			"InstanceName": name,
+			"tcp":          fmt.Sprint(binary.LittleEndian.Uint16(msg[4:6])),
```
Comment (chlafreniere) [inline tds.go]: **AI-generated review finding (Low):** This accepts `TCP_DAC_PORT == 0`. `tcpDialer.ParseBrowserData` preserves that zero, and `resolveServerPort` later maps it to the ordinary default TDS port 1433, silently losing DAC semantics. Port 0 is not a valid DAC endpoint; please reject it here and add a regression test using a validly shaped response whose port bytes are zero.

File: tds.go (hunk 2 above)
Comment (saurabh500) [reply tds.go]: Good catch — confirmed against the spec and fixed in d78f7de.

MC-SQLR 2.2.4 defines CLNT_UCAST_DAC as `0x0F`, PROTOCOLVERSION `0x01`, then a null-terminated instance name, and the worked example in 4.3 is:

```
Request:  0f 01 59 55 4b 4f 4e 53 54 44 00     | ..YUKONSTD
```

So the name starts at offset 2 and the packet ends with the terminator. `copy(bmsg[3:], instance)` was putting `0f 01 00 YUKONSTD` on the wire — a stray NUL before the name and no terminator after it. Worth noting the existing `make([]byte, 3+len(instance))` only fits the correct layout (2 header bytes + name + terminator), so the allocation was already right and only the offset was off.

Changed to `copy(bmsg[2:], instance)`, which leaves the zero-initialized final byte as the terminator.

Added `TestGetInstancesDACRequest`, which captures the datagram at a loopback UDP responder and asserts the exact bytes against the 4.3 vector. It fails on the previous offset:

```
- 00000000  0f 01 59 55 4b 4f 4e 53  54 44 00     |..YUKONSTD.|   (expected)
+ 00000000  0f 01 00 59 55 4b 4f 4e  53 54 44     |...YUKONSTD|   (before the fix)
```


> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: tds.go (hunk 3 above)
Comment (saurabh500) [reply tds.go]: Confirmed and fixed in d78f7de.

Traced the path: `ParseBrowserData` does `strconv.ParseUint("0", 0, 16)`, which succeeds, and assigns `p.Port = 0` (protocol.go:43-48). `resolveServerPort` then maps `0` to `defaultServerPort` (protocol.go:163-169) on the way into `DialSqlConnection`. So a validly shaped reply carrying port 0 produced a plain connection to 1433 while the caller believed it had reached the DAC endpoint — DAC semantics lost with no error.

Port 0 is never a listening endpoint, so `parseDAC` now rejects it rather than emitting an entry.

Added a `zero port` case to `TestParseDAC` and `TestParseDACZeroPortNotDefaulted`, which asserts the reply is refused, `Config.Port` is left unset, and documents the `resolveServerPort(0) == 1433` fallback that makes the rejection necessary.

For the record on threat model: this is correctness hygiene rather than a security boundary, since someone able to answer the UDP query could return any port they liked. The value is that a malformed or unusual reply now fails loudly instead of silently downgrading to the regular TDS endpoint.


> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

**Review by chlafreniere — APPROVED** [review]
